---
title: "ChIP Seq Annotation Script"
author: "Avantika Diwadkar (avantika.diwadkar@pennmedicine.upenn.edu)"
output:
  html_document:
    toc: TRUE
    depth: 3
editor_options: 
chunk_output_type: console
---
***

This report carries out annotation and visualization of differentially bound peaks/sites identified by processing ChIP-Seq data from GEO study. The steps involved are:

* Distribution of peak counts across the TSS region
* Annotation of the enriched regions
* Visualization of differentially bound sites to compare distance from TSS, annotations and overlap

Input:

A pre-prepared output file with a GRanges object saved from running ChIP_Seq_Diff_Binding_Analysis.Rmd script.

The input columns should show:

* the mean read concentration over all the samples: the default calculation uses log2 normalized ChIP read counts with control read counts subtracted
* the mean concentration of the first (dex) group
* the mean concentration of second (control) group
* the Fold column shows the difference in mean concentrations between the two groups (Conc_dex - Conc_control): a positive value indicating increased binding affinity in the dex group and a negative value indicating increased binding affinity in the control group.
* a raw p-value: confidence measures for identifying these sites as differentially
bound
* FDR: a multiple testing corrected FDR in the final column.

Outputs:

* Gene Annotation of the GRanges input file(s) (.bed/.txt)
* A ChIP-Seq annotation and visualization report (.html)


Manually change the variables for files (file1,file2) and conditions as well as txdb depending upon the genome used
```{r var_set, eval=T, echo=F, message=F, warning=F }
file1 = "/Users/diwadkar/Desktop/ChIP-Seq/GR_DB_sites.txt"
file2 = "/Users/diwadkar/Desktop/ChIP-Seq/RP_DB_sites.txt"
files = c(file1,file2)
conditions = c("GR","RNAP2")
```

Install the prerequisite R packages if they do not exist

* ChIPseeker
* TxDb.Hsapiens.UCSC.hg38.knownGene
* clusterProfiler
* org.Hs.eg.db
* dplyr and plyr
* pander

```{r pckgs, eval=F, echo=F, message=F, warning=F}
source("https://bioconductor.org/biocLite.R")
biocLite("ChIPseeker")
biocLite("TxDb.Hsapiens.UCSC.hg38.knownGene")
biocLite("clusterProfiler")
biocLite("org.Hs.eg.db")
biocLite("ReactomePA")
```

Loading the required libraries

```{r lib, eval=T, echo=F, message=F, warning=F}
library(ChIPseeker)
library(pander)
library(plyr)
library(dplyr)
library(ggplot2)
library(DT)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(clusterProfiler)
library(org.Hs.eg.db)
library(GenomicRanges)
library(ReactomePA)
```


####PRE-ANNOTATION ANALYSIS AND VISUALIZATION 

Read-in  and view your input files
```{r read_file, eval=T, echo=F, message=F, warning=F}
peak1 <- readPeakFile(file1[[1]])
peak2 <- readPeakFile(file2[[1]])
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
#seqlevels(txdb) <- seqlevels(txdb)[1:24]
nos <-  unlist(list(seq(1:22),"X","Y"))
chrs <- unlist(lapply(nos, function(x) paste0("chr",x)))

#GR results
head(as.data.frame(peak1))

#RNAP2 results
head(as.data.frame(peak2))
```

####Distribution of peaks across all chromosomes in the genome

The following plots show number of peaks in the two conditions and the distribution of peaks across chromosomes 1-22 and the sex chromosomes X and Y. 

The comparison between the GR and RNAP2 peak distributions shows denser or more peaks for GR than RNAP2. This is consistent with the number of peaks seen for GR (39k) which are much more than RNAP2(1.6k). 

Thus, you can visualize peaks across all chromosomes for the different conditions and confirm the result with number of peaks calculated prior. 

```{r peak_plot, eval=T, echo=F, message=F, warning=F}
df <- data.frame(Ab=conditions,Peaks = c(length(peak1),length(peak2)))
g <- ggplot(df, aes(x=Ab,y=Peaks,fill=Ab)) + geom_bar(stat = "identity",width = 0.3) + geom_text(aes(label = Peaks),position = position_dodge(width = 1),vjust=-0.5,size = 3) 
g <- g + theme_bw() + ggtitle("Peak count distribution for the two conditions") + labs(y="Peak counts",x="Antibody type") 
g

covplot(peak1, weightCol = "Conc",chrs = chrs,title = "ChIP Peaks for GR binding over Chromosomes")
covplot(peak2, weightCol = "Conc",chrs=chrs,title = "ChIP Peaks for RNAP2 binding over Chromosomes")
```

####ANNOTATION OF ENRICHED REGIONS IN THE DATASET

Peak Annotation is performed by annotatePeak where by default TSS(transcription start site) is defined from -3kb to +3kb. The output of annotatePeak is a csAnno instance. ChIPseeker provides as.GRanges to convert csAnno to GRanges instance, and  as.data.frame to convert csAnno to data.frame which can be exported to file by  write.table.TxDb object contained transcript-related features of a particular genome.

The genomic region of the peak is reported in the annotation column where the regions are defined as:
* Promoter
* 5’ UTR
* 3’ UTR
* Exon
* Intron
* Downstream
* Intergenic

Annotation of Peaks for Condition 1: GR binding

```{r annotate_gr, eval=T, echo=F, message=F, warning=F}
peakAnno1 <- annotatePeak(peak=file1[[1]], tssRegion = c(-3000,3000),TxDb = txdb,annoDb = "org.Hs.eg.db")
head(as.data.frame(peakAnno1))

peakAnno.one.db <- as.GRanges(peakAnno1)
write.table(peakAnno.one.db,paste0("Peaks.",conditions[1],".Annotated.txt"),sep="\t",row.names = FALSE)

plotAnnoPie(peakAnno1)
vennpie(peakAnno1)
upsetplot(peakAnno1)

#plotAnnoBar(peakAnno1)
#plotDistToTSS(peakAnno1)
```


Annotation of Peak for Condition 2: RNAP2 binding

```{r annotate_rp, eval=T, echo=F, message=F, warning=F}
peakAnno2 <- annotatePeak(peak=file2[[1]], tssRegion = c(-3000,3000),TxDb = txdb,annoDb = "org.Hs.eg.db")
head(as.data.frame(peakAnno2))

peakAnno.two.db <- as.GRanges(peakAnno2)
write.table(peakAnno.two.db,paste0("Peaks.",conditions[2],".Annotated.txt"),sep="\t",row.names = FALSE)

plotAnnoPie(peakAnno2)
vennpie(peakAnno2)
upsetplot(peakAnno2)
#plotAnnoBar(peakAnno2)
#plotDistToTSS(peakAnno2)
```

####Functional enrichment analysis

Enrichment analysis is a widely used approach to identify biological themes and investigate whether the number 
of selected genes associated with a particular biological term is larger than expected, using packages like DOSE for Disease Ontology, 
ReactomePA for reactome pathway, clusterProfiler for Gene Ontology and KEGG enrichment analysis.

```{r enrich_GO, eval=T, echo=F, message=F, warning=F}
bp1 <- enrichGO(unlist(peakAnno.one.db$geneId), OrgDb="org.Hs.eg.db", ont = "BP",readable = TRUE)
bp1_df <- as.data.frame(bp1)
bp1_df$geneID <- NULL
head(bp1_df)

bp2 <- enrichGO(unlist(peakAnno.two.db$geneId), OrgDb="org.Hs.eg.db", ont = "BP",readable = TRUE)
bp2_df <- as.data.frame(bp2)
bp2_df$geneID <- NULL
head(bp2_df)

```

```{r reactome_pa, eval=T, echo=F, message=F, warning=F}
gene1 <- seq2gene(peak1, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
pathway1 <- enrichPathway(gene1)
dotplot(pathway1,title = "Enrichment result of annotated GR sites")


gene2 <- seq2gene(peak2, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
pathway2 <- enrichPathway(gene2)
dotplot(pathway2,title = "Enrichment result of annotated RNAP2 sites")
```



####COMPARISON OF DIFFERENTIALLY BOUND ENRICHED SITES

This section compares annotations and distribution of peaks idenitfied in the two conditions. Input for this is both files together. 

The plot and heatmap looks at the distribution of peak count frequency from TSS of the genes in upstream direction for the two conditions.

```{r tagMatrix, eval=T, echo=F, message=F, warning=F}
promoter <- getPromoters(TxDb=txdb,upstream=3000, downstream=3000) 
tagMatrixList <- lapply(files, getTagMatrix, windows=promoter)
tagMatrixList <- setNames(tagMatrixList, conditions)
plotAvgProf(tagMatrixList, xlim = c(-3000, 3000))

tagHeatmap(tagMatrixList, xlim = c(-3000, 3000), color = NULL)
```


####BARPLOTS

The first bar plot look at annotation differences in feature distribution of the two conditions where features include promotor regions, UTRs, exons, introns, and distal intergenic regions. 

The second bar plot looks at percentage distribution of TF-binding loci position relative to TSS of the genes.

```{r peakAnno, eval=T, echo=F, message=F, warning=F}
peakAnnoList <- lapply(files, annotatePeak, TxDb = txdb, tssRegion = c(-3000, 3000), verbose = FALSE)
peakAnnoList <- setNames(peakAnnoList, conditions)
plotAnnoBar(peakAnnoList)
plotDistToTSS(peakAnnoList)
```

####FUNCTIONAL PROFILE COMPARISON

The ReactomePA package provides a function, seq2gene, for linking genomc regions to genes in a many-to-many mapping. It consider host gene (exon/intron), promoter region and flanking gene from intergenic region that may under control via cis-regulation. 

This function is designed to link both coding and non-coding genomic regions to coding genes and facilitate functional analysis.

```{r reactome_pa2, eval=T, echo=F, message=F, warning=F}
all_genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
names(all_genes) = sub("_", "\n", names(all_genes))
compKEGG <- compareCluster(geneCluster   = all_genes,
                         fun           = "enrichKEGG",
                         pvalueCutoff  = 0.05,
                         pAdjustMethod = "BH")
dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


####VENNPLOT

Venn Diagram looks into overlap of genes annotated to enriched regions in the two conditions with fold change greater than 2. The table shows all the overlapped genes. 

```{r venn,eval=T, echo=F, message=F, warning=F}
#genes1 <- (as.data.frame(peakAnno.one.db) %>% filter(abs(Fold) > 2))$SYMBOL
#genes2 <- (as.data.frame(peakAnno.two.db) %>% filter(abs(Fold) > 2))$SYMBOL
genes1 <- (as.data.frame(peakAnno.one.db))$SYMBOL
genes2 <- (as.data.frame(peakAnno.two.db))$SYMBOL           
genes = list(genes1, genes2)
genes = setNames(genes, conditions)
vennplot(genes)
overlap_genes <- data.frame(Genes=intersect(genes1,genes2))
DT::datatable(overlap_genes, rownames=FALSE, options = list(columnDefs = list(list(className = 'dt-center', targets = "_all"))))

#Get conc for specific genes in both conditions
GR_df <- as.data.frame(peakAnno.one.db) 
col_names <- names(GR_df)
GR_df[col_names] <- lapply(GR_df[col_names] , factor)
GR_df <-  GR_df %>% dplyr::select(Fold,SYMBOL) %>% filter(SYMBOL %in% overlap_genes$Genes)
GR_df$p.value <- NULL

RP_df <- as.data.frame(peakAnno.two.db) 
col_names <- names(RP_df)
RP_df[col_names] <- lapply(RP_df[col_names] , factor)
RP_df <-  RP_df %>% dplyr::select(Fold,SYMBOL) %>% filter(SYMBOL %in% overlap_genes$Genes)
RP_df$p.value <- NULL

main_df <- merge(GR_df, RP_df, by="SYMBOL")
colnames(main_df) <- c("SYMBOL","GR","RNAP2")
```

#### Session information

```{r sessioninfo, eval=T, echo=F}
pander(sessionInfo())
```
