brocade
======

Reproducible analysis of ChIP-Seq data

Authors: Mengyuan Kan, Avantika Diwadkar, Blanca Himes

## Overview

The goal of [brocade pipeline|@https://github.com/HimesGroup/brocade@] is to preform reproducible analysis of ChIP-Seq data:

* Download SRA .fastq data and prepare phenotype file
* Align reads to a reference genome
* Call peaks enriched in protein-DNA binding regsions
* Perform QC on aligned files
* Create a report that can be used to verify that sequencing was successful and/or identify sample outliers
* Perform differential binding analysis of reads aligned to genome according to a given reference genome
* Annotate genomic features to peaks
* Create a report that summarizes the differential binding results
Automatically generate LSF scripts in each step for HPC use.

## Bioinformatics Tools

* **Alignment:** BWA
* **Peak Calling:** macs2
* **QC:** fastqc, trimmomatic, samtools, bamtools, picardtools
* **Differential binding analysis:** DiffBind
* **Peak genomic feature annotation:** Corresponding R genome database packages and ChIPseeker
* **Pipeline scripts:** the Python scripts make use of various modules including subprocess, os, argparse, sys
**R markdown scripts for summary reporot:** Require various R libraries such as DT, gplots, ggplot2, rmarkdown, RColorBrewer, dplyr. Note that the current RMD scripts require pandoc version 1.12.3 or higher to generate HTML report.

## Prerequisite Files

* **Spesis-specific genome reference files:** reference genome fasta files (.fa) and corresponding BWA index files.
* **Adapter and primer sequences**: a list of adapter and primer sequences is provided in **template_files/adapter_primer_sequences.txt**. For fastqc, replace . For trimming we provide adapter and primer sequences for the following types: Ilumina TruSeq single index, Illumina unique dual (UD) index adpter and PrepX. Users can tailor this file by adding sequences from other protocols.

## Workflow

### Download data from GEO/SRA

Run **pipeline_scripts/chipseq\_sra\_download.py** to download .fastq files from SRA. Read in **template_files/chipseq_sra_download_Rmd_template.txt** from specified directory <i>template_dir</i> to create a RMD script. Ftp addresses for corresponding samples are obtained from SRA SQLite database using R package SRAdb.

> chipseq\_sra\_download.py --geo\_id <i>GEO_Accession</i> --path\_start <i>output_path</i> --project\_name <i>output_prefix</i> --template\_dir <i>templete_file_directory</i> --fastqc

> bsub < <i>project_name</i>_download.lsf

The option *--pheno_info* refers to using user provided SRA ID for download which is included in the SRA_ID column in the provided phenotype file. If the phenotype file is not provided, use phenotype information from GEO. SRA_ID is retrieved from the field relation.1.

The option *--fastqc* refers to running FastQC for downloaded .fastq files.

**Output files:** 

Fastq files downloaded from SRA are saved in the following directory.

> <i>path_start</i>/<i>project_name</i>_SRAdownload/

Fastqc results of corresponding sample are saved in:

> <i>path_start</i>/<i>sample_name</i>/<i>fastq_name</i>_fastqc.zip

The RMD and corresponding HTML report files:

> <i>path_start</i>/<i>project_name</i>_SRAdownload/<i>project_name</i>__SRAdownload_ChIPSeqReport.Rmd
> <i>path_start</i>/<i>project_name</i>_SRAdownload/<i>project_name</i>__SRAdownload\_ChIPSeqReport.html

GEO phenotype file generated by RMD reports:

> <i>path_start</i>/<i>project_name</i>_SRAdownload/<i>GEO_Accession</i>_GEO_phenotype.txt

SRA information file generated by RMD reports:

> <i>path_start</i>/<i>project_name</i>_SRAdownload/<i>project_name</i>_sraFile.info

### User-tailored phentoype file

The sample info file used in the following steps should be provided by users.

**Required columns:**

* 'Sample' column containing sample ID
* 'Status' column containing variables of comparison state
* 'Antibody' column containing antibody used for ChIP
* 'Tissue' column containing tissue used
* 'Treatment' column containing treatment conditions
* 'Input' column containing sample ID used for DNA intput control that did not undergo ChIPSeq
* 'Peak' column specify 'narrow' or 'broad' peak type
* 'R1' and/or 'R2' columns containing full paths of .fastq files

**Other columns:**

'Disease', 'Donor' (i.e. cell line ID if <i>in vitro</i> treatment is used), 'Tissue', 'protocol' designating sample preparation kit information.

**'Index' column** contains index sequence for each sample. If provided, trim raw .fastq files based on corresponding adapter sequences.

If use data from GEO, most GEO phenotype data do not have index information. However, FastQC is able to detect them as "Overrepresented sequences". Users can tailor the 'Index' column based on FastQC results. We provide a file **template_files/adapter_primer_sequences.txt** with most updated adapter and primer sequences for FastQC detection.

In this example, we used a customized adapter file **example_files/adapter_primer_sequences.txt**.

An example phenotype file can be found here: **example_files/GSE95632_Phenotype_withoutQC.txt**. Note that column naming is rigid for the following columns: 'Sample', 'Status', 'Antibody', 'Tissue', 'Treatment', 'Input', 'Index', 'R1', 'R2', 'Disease', 'Donor', because pipeline scripts will recognize these name strings, but the column order can be changed.

### Alignment, peak calling and QC

1) Run **pipeline_scripts/chipseq_align_and_qc.py** to: 1) trim adapter and primer sequences if index information is available, 2) run FastQC for (un)trimmed .fastq files, 3) align reads to reference genome, and 5) obtain various QC metrics from .bam files.

Edit **pipeline_scripts/chipseq_userdefine_variables.py** with a list user-defined variables (e.g. paths of genome reference file, paths of bioinformatics tools, versions of bioinformatics tools), and save the file under an executable search path.

If perform adapter trimming, read in **template_files/adapter_primer_sequences.txt** from specified directory <i>template_dir</i> used as a reference list of index and primer sequences for various library preparation kits. In this example, a customized adapter file **example_files/adapter_primer_sequences.txt** was used. To replicate this result, save this file **example_files/adapter_primer_sequences.txt** in user-defined <i>templete_file_directory</i>.

> chipseq_align_and_qc.py --project_name <i>output_prefix</i> --samples_in <i>sample_info_file.txt</i> --ref_genome hg38 --library_type SE --index_type GSE95632 --path_start <i>output_path</i> --template\_dir <i>templete_file_directory</i>

> for i in *align.lsf; do bsub < $i; done

The "--ref\_genome" option refers to using selected version of genome reference.

The "--library\_type" option refers to PE (paired-end) or SE (single-end) library.

The "--index\_type" option refers to index used in sample library preparation.

Common index types provided in **template_files/chipseq_adapter_primer_sequences.txt** are:

* truseq\_single\_index (TruSeq Single Indexes)
* illumina\_ud\_sys1 (Illumina UD indexes for NovaSeq, MiSeq, HiSeq 2000/2500)
* illumina\_ud\_sys2 (Illumina UD indexed for MiniSeq, NextSeq, HiSeq 3000/4000)
* prepX (PrepX for Apollo 324 NGS Library Prep System)

**template_files/chipaseq_adapter_primer_sequences.txt** contains four columns (i.e. Type, Index, Description, Sequence). Sequences in the Index column is used to match those in Index column in sample info file. This column naming is rigid.

The list is based on the following resources:

* [Illumina adapter sequences](https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences-1000000002694-07.pdf)
* [PrepX RNA-Seq Index Primers and Sequences](https://genome.med.harvard.edu/documents/illumina/IntegenX_Apollo324_mRNA_Seq_Protocol_10012012.pdf)

If users provide new sequences, add the new index type in the 1st column 'Type' and specify it in "--index\_type".

The "--strand" option refers to sequencing that captures both strands (nonstrand) or the 1st synthesized strand (reverse) or the 2nd synthesized strand (forward) of cDNA. If the 2nd strand is synthesized using dUTP, this strand will extinct during PCA amplification, thus only 1st (reverse) strand will be sequenced.

Read sample preparation protocal carefully. Reads not in the specified strand will be discarded. Double check proprotion of reads mapped to no feature category in QC report. If a lot of reads are mapped to 'no feature', the strand option setting is likely incorrect.

**Output files:**

Various output files will be written for each sample in directories structured as:

> <i>path_start</i>/<i>sample_name</i>/<i>sample_name</i>_R1_Trimmed.fastq <br>
> <i>path_start</i>/<i>sample_name</i>/<i>sample_name</i>_R2_Trimmed.fastq <br>
> <i>path_start</i>/<i>sample_name</i>/<i>sample_name</i>_R1_Trimmed\_fastqc.zip <br>
> <i>path_start</i>/<i>sample_name</i>/<i>sample_name</i>_R2_Trimmed\_fastqc.zip <br>
> <i>path_start</i>/<i>sample_name</i>/<i>sample_name</i>_ReadCount <br>
> <i>path_start</i>/<i>sample_name</i>/bwa\_out <br>

2) Run **pipeline_scripts/chipseq\_align\_and\_qc\_report.py** to call peaks.

> chipseq\_peakcaller.py --project\_name <i>output_prefix</i> --samples\_in <i>output_prefix</i> --ref\_genome hg38 --path\_start <i>output_path</i> --template\_dir <i>templete_file_directory</i>

> for i in *macs2.lsf; do bsub < $i; done

Filtering out peaks within 'blacklist regions'. Blacklist regions for hg38 and hg19 are provided in **templete_files**: **hg19_blacklist.bed** and **hg38_blacklist.bed**.

**Output files:**

Peak files will be written for each sample that underwent ChIPSeq in the macs2 output directory:
> <i>path_start</i>/<i>sample_name</i>/macs2\_out <br>

3) Run **pipeline_scripts/chipseq\_align\_and\_qc\_report.py** to create an HTML report of QC and alignment summary statistics for ChIP-Seq samples. Read in **template\_files/chipseq\_align\_and\_qc\_report\_Rmd\_template.txt** from specified directory <i>template_dir</i> to create a RMD script.

> chipseq\_align\_and\_qc\_report.py  --project\_name <i>output_prefix</i> --sample\_in <i>sample_info_file.txt</i> --ref_genome hg38 --library_type SE --path\_start <i>output_path</i> --template\_dir <i>templete_file_directory</i>

> bsub < <i>project_name</i>_qc.lsf

**Output files:**

This script uses the many output files created in step 1), converts these sample-specific files into matrices that include data for all samples, and then creates an Rmd document.

The report and accompanying files are contained in:

> <i>path_start</i>/<i>project_name</i>_Alignment_QC\_Report/

The RMD and corresponding HTML report files:

> <i>path_start</i>/<i>project_name</i>_Alignment_QC\_Report/<i>project_name</i>_QC_ChIPSeqReport.Rmd
> <i>path_start</i>/<i>project_name</i>_Alignment_QC\_Report/<i>project_name</i>_QC_ChIPSeqReport.html

### Differential binding analysis and genomic feature annotation

Run **pipeline\_scripts/chipseq\_diffbind\_report.py** to perform differential binding analysis and create an HTML report of differential binding summary statistics. Read in **template\_files/chipseq\_diffbind\_report\_Rmd\_template.txt** from specified directory <i>template_dir</i> to create a RMD script.

> chipseq_diffbind_report.py --project_name <i>output_prefix</i> --samples_in <i>sample_info_file_withQC.txt</i> --comp <i>sample_comp_file.txt</i> --ref_genome hg38 --path_start <i>output_path</i> --template_dir <i>templete_file_directory</i>


> bsub < <i>project_name</i>_diffbind.lsf

The "--sample_in" option specifies user provided phenotype file for DE analysis (**example\_files/GSE95632_Phenotype_withQC.txt**). The columns are the same as **example\_files/GSE95632_Phenotype_withoutQC.txt** but with an additional column "QC_Pass" designating samples to be included (QC_Pass=1) or excluded (QC_Pass=0) after QC. This column naming is rigid which will be recoganized in pipeline scripts, but column order can be changed.

The "--comp" option specifies comparisons of interest in a tab-delimited text file with one comparison per line with three columns (i.e. Condition1, Condition0, Design), designating Condition1 vs. Condition2. The current version of differential analysis does not support paired analysis.

Find the example comp file here **example\_files/GSE95632\_comp\_file.txt**.

**Output files:**

Files generated for this step are contained in:

> <i>project_name</i>/<i>project_name</i>_diffbind_out/

The sample sheet .csv files used as DiffBind input:

> <i>project_name</i>_<i>Condition1</i>\_vs\_<i>Condition2</i>\.sampleinfo.csv

The pairwise differential binding results and normalized counts for significant binding sites:

> <i>project_name</i>_<i>Condition1</i>\_vs\_<i>Condition2</i>\_sig_diffbind_results.csv
> <i>project_name</i>_<i>Condition1</i>\_vs\_<i>Condition2</i>\_sig_counts_normalized_by_diffbind.csv

The RMD and corresponding HTML report file:

> <i>path_start</i>/<i>project_name</i>_deseq2_out/<i>project_name</i>_DiffBind_Report.Rmd
> <i>path_start</i>/<i>project_name</i>_deseq2_out/<i>project_name</i>_DiffBind_Report.html
